src/rules/store/selectors-start-with-select.ts

```diff
@@ -0,0 +1,36 @@
+import path from 'path'
+import { ESLintUtils, TSESTree } from '@typescript-eslint/experimental-utils'
+
+import { docsUrl } from '../../utils'
+
+export const messageId = 'selectInSelectorNames'
+export type MessageIds = typeof messageId
+
+type Options = []
+
+export default ESLintUtils.RuleCreator(docsUrl)<Options, MessageIds>({
+  name: path.parse(__filename).name,
+  meta: {
+    type: 'suggestion',
+    docs: {
+      category: 'Best Practices',
+      description:
+        'The variable name of a selector should start with "select", for example "selectThing"',
+      recommended: 'warn',
+    },
+    schema: [],
+    messages: {
+      [messageId]: 'The variable name of a selector should start with "select"',
+    },
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      'VariableDeclarator[id.name=/^(?!select).+/]:matches(:has(TSTypeAnnotation[typeAnnotation.typeName.name=/^MemoizedSelector(Props$|$)/]), :has(CallExpression[callee.name=/^(createSelector|createFeatureSelector)$/]))'({
```

Comment:
After a second look to this, I noticed two things here:

1. What should happen if one uses, for example, `select` or `selectxx` words? I'm asking because the current pattern treats this as valid... **but** shouldn't we ensure there's an uppercase after the 'select' word? Maybe it could impact the rule name too, as start with !== prefixed with, no? Anyway, let me know your opinions about this (it'd be nice to have test cases for this).

2. It should be `MemoizedSelector(WithProps$|$)`, not `MemoizedSelector(Props$|$)` (it's missing the 'With'). However, is it something we should care about? It's currently **deprecated** as documented in https://ngrx.io/api/store/MemoizedSelectorWithProps. If we decide to care about it, it'd be nice to have a test case for this then.

cc @timdeschryver

---

Here's the version that requires 'select[A-Z]' pattern and **with** `MemoizedSelectorWithProps`:

 ```suggestion
       'VariableDeclarator[id.name!=/^(^select)[A-Z][a-zA-Z]+$/]:matches(:has(TSTypeAnnotation[typeAnnotation.typeName.name=/^MemoizedSelector(WithProps$|$)/]), :has(CallExpression[callee.name=/^(createSelector|createFeatureSelector)$/]))'({
```

And here's the version that requires 'select[A-Z]' pattern and without `MemoizedSelectorWithProps`:

 ```suggestion
       'VariableDeclarator[id.name!=/^(^select)[A-Z][a-zA-Z]+$/]:matches(:has(TSTypeAnnotation[typeAnnotation.typeName.name=/^MemoizedSelector$/]), :has(CallExpression[callee.name=/^(createSelector|createFeatureSelector)$/]))'({
```

Reply:
I think we should do 1 (the casing) because typos are quickly made 😅
I wouldn't bother with 2 (with props), and just leave it as is 

The second suggested change seems good to me

Reply:
Nice, @timdeschryver. In this case should we still call it `selectors-start-with-select` or would it be better to call it `selectors-prefix-with-select`?

@prince-chrismc I just noticed that you opted for maintaining `MemoizedSelectorWithProps`... so in this case, it'd be good to have examples in `.md` and test cases (a success and a fail case should be good). Also, a example + test case for the word "select" 😆

Reply:
That was not intentional, however we still use props and this rule does not cover the `createSelectorFactory` which I believe is the replacement?

Reply:
I think `prefix-selectors-with-select` would be the best then if both of you agree.
The replacement is a selector that returns a selector, e.g.

```ts
const selectThingy = (id: string) => createSelector(selectThings, things => things[id]);
```

I think for now, this rule covers like 90% of the cases and we can merge this in if we agree on the name.
Later, we can follow-up with a PR to address the factory selectors.

Reply:
I agree! 👍